### PR TITLE
[HINTS] Div mod n hint

### PR DESCRIPTION
**Pull request by Juan-M-V**
# Div mod n hint

## Description

Must be merged before #338
Implements div_mod_n from signature

## Checklist
- [x] Unit tests added
- [x] Integration tests added.

### CODE REVIEW

File: src/math_utils.rs [hunk 1]
```diff
@@ -25,6 +25,15 @@ pub fn isqrt(n: &BigInt) -> Result<BigInt, VirtualMachineError> {
     Ok(x)
 }
 
+/// Performs integer division between x and y; fails if x is not divisible by y.
+pub fn safe_div(x: &BigInt, y: &BigInt) -> Result<BigInt, VirtualMachineError> {
+    if !(x % y).is_zero() || y.is_zero() {
```
Comment (Oppen) [inline src/math_utils.rs:30]: You need to invert this check, otherwise you risk dividing by zero.
```suggestion
    if y.is_zero() || !(x % y).is_zero() {
```

File: src/math_utils.rs [hunk 2]
```diff
@@ -69,10 +78,10 @@ fn igcdex(num_a: BigInt, num_b: BigInt) -> (BigInt, BigInt, BigInt) {
     }
 }
 ///Finds a nonnegative integer x < p such that (m * x) % p == n.
-fn div_mod(n: BigInt, m: BigInt, p: BigInt) -> BigInt {
+pub fn div_mod(n: BigInt, m: BigInt, p: &BigInt) -> BigInt {
```
Comment (Oppen) [inline src/math_utils.rs:81]: `n` and `b` should probably be references as well.

File: src/math_utils.rs [hunk 3]
```diff
@@ -92,7 +101,7 @@ pub fn ec_add(
 /// Assumes the points are given in affine form (x, y) and have different x coordinates.
 pub fn line_slope(point_a: (BigInt, BigInt), point_b: (BigInt, BigInt), prime: &BigInt) -> BigInt {
```
Comment (Oppen) [inline src/math_utils.rs:102]: An the same for `point_a` and `point_b`.

File: src/vm/hints/hint_utils.rs [hunk 4]
```diff
@@ -28,6 +28,19 @@ pub fn get_int_from_scope(vm: &mut VirtualMachine, name: &str) -> Option<BigInt>
     val
 }
 
+pub fn get_int_from_scope_ref<'a>(
+    vm: &'a mut VirtualMachine,
```
Comment (Oppen) [inline src/vm/hints/hint_utils.rs:32]: Does this need to be mutable?

File: src/vm/hints/secp/field_utils.rs [hunk 5]
```diff
@@ -72,18 +66,9 @@ pub fn reduce(
...
-
-    let x_d0 = get_integer_from_relocatable_plus_offset(&x_reloc, 0, vm)?;
-    let x_d1 = get_integer_from_relocatable_plus_offset(&x_reloc, 1, vm)?;
-    let x_d2 = get_integer_from_relocatable_plus_offset(&x_reloc, 2, vm)?;
-
-    //SECP_P = 2**256 - 2**32 - 2**9 - 2**8 - 2**7 - 2**6 - 2**4 - 1
-    let sec_p = bigint_str!(
-        b"115792089237316195423570985008687907853269984665640564039457584007908834671663"
-    );
+    let value = pack_from_var_name("x", ids, vm, hint_ap_tracking)?.mod_floor(&*SECP_P);
```
Comment (Oppen) [inline src/vm/hints/secp/field_utils.rs:69]: This looks weird. If we have a reference and we pass a reference, why do we need to dereference first?

File: src/vm/hints/secp/field_utils.rs [hunk 6]
```diff
@@ -72,18 +66,9 @@ pub fn reduce(
...
 
-    let value = pack(x_d0, x_d1, x_d2, &vm.prime).mod_floor(&sec_p);
+    //let value = pack(x_d0, x_d1, x_d2, &vm.prime).mod_floor(&sec_p);
```
Comment (Oppen) [inline src/vm/hints/secp/field_utils.rs:71]: Try not to leave commented out code.

File: src/vm/hints/secp/secp_utils.rs [hunk 7]
```diff
@@ -48,6 +61,21 @@ pub fn pack(d0: &BigInt, d1: &BigInt, d2: &BigInt, prime: &BigInt) -> BigInt {
         .sum()
 }
 
+pub fn pack_from_var_name(
+    name: &str,
+    ids: &HashMap<String, BigInt>,
+    vm: &mut VirtualMachine,
```
Comment (Oppen) [inline src/vm/hints/secp/secp_utils.rs:67]: Do we need this to be mutable?

File: src/vm/hints/secp/signature.rs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+// Implements hint:
+// value = k = safe_div(res * b - a, N)
+pub fn div_mod_n_safe_div(vm: &mut VirtualMachine) -> Result<(), VirtualMachineError> {
+    let a = get_int_from_scope_ref(vm, "a")?.clone();
+    let b = get_int_from_scope_ref(vm, "b")?.clone();
+    let res = get_int_from_scope_ref(vm, "res")?;
+
+    let value = safe_div(&(res * b - a), &*N)?;
```
Comment (Oppen) [inline src/vm/hints/secp/signature.rs:53]: Can't the reference to `N` be passed directly?

File: src/vm/hints/secp/signature.rs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        vm.exec_scopes
+            .assign_or_update_variable("a", PyValueType::BigInt(bigint!(0_usize)));
+        vm.exec_scopes
+            .assign_or_update_variable("b", PyValueType::BigInt(bigint!(1_usize)));
+        vm.exec_scopes
+            .assign_or_update_variable("res", PyValueType::BigInt(bigint!(1_usize)));
+
+        assert_eq!(Err(VirtualMachineError::SafeDivFail(bigint!(1_usize), bigint_str!(b"115792089237316195423570985008687907852837564279074904382605163141518161494337"))), div_mod_n_safe_div(&mut vm));
+    }
+}
```
Comment (Oppen) [inline src/vm/hints/secp/signature.rs:82]: Don't we also need a test for the passing case?

File: src/math_utils.rs [hunk 10]
```diff
@@ -185,7 +194,24 @@ mod tests {
...
+        assert_eq!(safe_div(&x, &y), Ok(bigint!(2)));
+    }
+
+    #[test]
+    fn compute_safe_div_fail() {
+        let x = bigint!(25);
+        let y = bigint!(4);
+        assert_eq!(
+            safe_div(&x, &y),
+            Err(VirtualMachineError::SafeDivFail(bigint!(25), bigint!(4)))
```
Comment (Oppen) [inline src/math_utils.rs:214]: Add a test case for division by zero.

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/math_utils.rs [hunk 11]
```diff
@@ -25,6 +25,15 @@ pub fn isqrt(n: &BigInt) -> Result<BigInt, VirtualMachineError> {
...
 }
 
+/// Performs integer division between x and y; fails if x is not divisible by y.
+pub fn safe_div(x: &BigInt, y: &BigInt) -> Result<BigInt, VirtualMachineError> {
+    if !(x % y).is_zero() || y.is_zero() {
+        Err(VirtualMachineError::SafeDivFail(x.clone(), y.clone()))
+    } else {
+        Ok(x / y)
+    }
+}
```
Comment (Oppen) [inline src/math_utils.rs:35]: In fact, you may want to also perform the division only once:
```suggestion
pub fn safe_div(x: &BigInt, y: &BigInt) -> Result<BigInt, VirtualMachineError> {
    if y.is_zero() {
        return Err(VirtualMachineError::DividedByZero);
    }
    let (q, r) = x.div_rem(y);
    if !r.is_zero() {
        return Err(VirtualMachineError::SafeDivFail(x.clone(), y.clone()));
    }
    Ok(q)
```

File: src/math_utils.rs (hunk 1 above)
Comment (Juan-M-V) [reply src/math_utils.rs:30]: Resolved: 5eed780ae84ab1dbd34da6e7d94f154b619f1200

File: src/vm/hints/secp/field_utils.rs (hunk 5 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/field_utils.rs:69]: That function converts a struct to a single BigInt value. In that case x is actually a struct with the values x.d0, x.d1 and x.d2

File: src/vm/hints/secp/signature.rs (hunk 8 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/signature.rs:53]: As far as I understand, this has to do with lazy_static and how it needs to be called. Something about doing the computation the first time its called

File: src/math_utils.rs (hunk 10 above)
Comment (Juan-M-V) [reply src/math_utils.rs:214]: I can't because the divisor is always N

File: src/math_utils.rs (hunk 11 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/bff8fe8a880eec8430a711e8508ccbaee3da8247

File: src/vm/hints/secp/signature.rs (hunk 9 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/signature.rs:132]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/06324ae488d4f319383f038995bcd2544b1f464a

File: src/math_utils.rs (hunk 10 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: partially resolved:
https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/bff8fe8a880eec8430a711e8508ccbaee3da8247

File: src/vm/hints/secp/field_utils.rs (hunk 6 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/field_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/ed307c54ee93a1c8994d030108ed33efbba98391

File: src/vm/hints/secp/secp_utils.rs (hunk 7 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/secp_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/f9552216b25097ece469f70cff35ee8620aa5b64

File: src/math_utils.rs (hunk 3 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: 
https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/fdcfbe283adf510aaf7cb1982f6c256f43f8ec64

File: src/math_utils.rs (hunk 2 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved:
- https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/db90895a67ed004b2280e763562e66168a408a7b
- https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/2e6aaad6adbd1f438678c587d5aa493fdebeb1c7

File: src/vm/hints/hint_utils.rs (hunk 4 above)
Comment (Juan-M-V) [reply src/vm/hints/hint_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/5a9bd80e9b1ee4158042e209384e3372f6e48792

File: src/math_utils.rs (hunk 1 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/5eed780ae84ab1dbd34da6e7d94f154b619f1200

File: src/math_utils.rs [hunk 12]
```diff
@@ -51,25 +57,30 @@ fn igcdex(num_a: BigInt, num_b: BigInt) -> (BigInt, BigInt, BigInt) {
...
     match (a.clone(), b.clone()) {
-        (a, b) if a == b && a == bigint!(0) => (bigint!(0), bigint!(1), bigint!(0)),
-        (a, _) if a == bigint!(0) => (bigint!(0), b.div_floor(&abs(b.clone())), abs(b)),
-        (_, b) if b == bigint!(0) => (a.div_floor(&a), bigint!(0), abs(a)),
+        (a, b) if a == b && a == bigint!(0_i32) => (bigint!(0_i32), bigint!(1_i32), bigint!(0_i32)),
+        (a, _) if a == bigint!(0_i32) => (bigint!(0_i32), b.div_floor(&abs(b.clone())), abs(b)),
+        (_, b) if b == bigint!(0_i32) => (a.div_floor(&a), bigint!(0_i32), abs(a)),
         _ => {
-            if a < bigint!(0) {
+            if a < bigint!(0_i32) {
```
Comment (Oppen) [inline src/math_utils.rs:64]:
```suggestion
            if a.is_negative() {
```

File: src/math_utils.rs [hunk 13]
```diff
@@ -51,25 +57,30 @@ fn igcdex(num_a: BigInt, num_b: BigInt) -> (BigInt, BigInt, BigInt) {
     let x_sign: i32;
     let y_sign: i32;
     match (a.clone(), b.clone()) {
-        (a, b) if a == b && a == bigint!(0) => (bigint!(0), bigint!(1), bigint!(0)),
-        (a, _) if a == bigint!(0) => (bigint!(0), b.div_floor(&abs(b.clone())), abs(b)),
-        (_, b) if b == bigint!(0) => (a.div_floor(&a), bigint!(0), abs(a)),
+        (a, b) if a == b && a == bigint!(0_i32) => (bigint!(0_i32), bigint!(1_i32), bigint!(0_i32)),
+        (a, _) if a == bigint!(0_i32) => (bigint!(0_i32), b.div_floor(&abs(b.clone())), abs(b)),
+        (_, b) if b == bigint!(0_i32) => (a.div_floor(&a), bigint!(0_i32), abs(a)),
```
Comment (Oppen) [inline src/math_utils.rs:62]:
```suggestion
        (a, b) if a == b && a.is_zero() => (bigint!(0_i32), bigint!(1_i32), bigint!(0_i32)),
        (a, _) if a.is_zero() => (bigint!(0_i32), b.div_floor(&abs(b.clone())), abs(b)),
        (_, b) if b.is_zero() => (a.div_floor(&a), bigint!(0_i32), abs(a)),
```

File: src/math_utils.rs [hunk 14]
```diff
@@ -51,25 +57,30 @@ fn igcdex(num_a: BigInt, num_b: BigInt) -> (BigInt, BigInt, BigInt) {
...
         _ => {
-            if a < bigint!(0) {
+            if a < bigint!(0_i32) {
                 a = -a;
                 x_sign = -1;
             } else {
                 x_sign = 1;
             }
-            if b < bigint!(0) {
+            if b < bigint!(0_i32) {
```
Comment (Oppen) [inline src/math_utils.rs:70]:
```suggestion
            if b.is_negative() {
```

File: src/math_utils.rs [hunk 15]
```diff
@@ -51,25 +57,30 @@ fn igcdex(num_a: BigInt, num_b: BigInt) -> (BigInt, BigInt, BigInt) {
...
-            let (mut x, mut y, mut r, mut s) = (bigint!(1), bigint!(0), bigint!(0), bigint!(1));
+            let (mut x, mut y, mut r, mut s) = (
+                bigint!(1_i32),
+                bigint!(0_i32),
+                bigint!(0_i32),
+                bigint!(1_i32),
+            );
             let (mut c, mut q);
-            while b != bigint!(0) {
+            while b != bigint!(0_i32) {
```
Comment (Oppen) [inline src/math_utils.rs:83]:
```suggestion
            while !b.is_zero() {
```

File: src/math_utils.rs [hunk 16]
```diff
@@ -91,17 +102,25 @@ pub fn ec_add(
...
 /// Assumes the points are given in affine form (x, y) and have different x coordinates.
-pub fn line_slope(point_a: (BigInt, BigInt), point_b: (BigInt, BigInt), prime: &BigInt) -> BigInt {
-    assert!((point_a.0.clone() - point_b.0.clone()) % prime != bigint!(0));
-    div_mod(point_a.1 - point_b.1, point_a.0 - point_b.0, prime)
+pub fn line_slope(
+    point_a: &(BigInt, BigInt),
+    point_b: &(BigInt, BigInt),
+    prime: &BigInt,
+) -> BigInt {
+    assert!((point_a.0.clone() - point_b.0.clone()) % prime != bigint!(0_i32));
```
Comment (Oppen) [inline src/math_utils.rs:118]:
```suggestion
    assert!(!((point_a.0.clone() - point_b.0.clone()) % prime).is_zero();
```

File: src/math_utils.rs [hunk 17]
```diff
@@ -117,10 +136,10 @@ pub fn ec_double(point: (BigInt, BigInt), alpha: &BigInt, prime: &BigInt) -> (Bi
 /// the given point.
 /// Assumes the point is given in affine form (x, y) and has y != 0.
 pub fn ec_double_slope(point: (BigInt, BigInt), alpha: &BigInt, prime: &BigInt) -> BigInt {
-    assert!(point.1.clone() % prime != bigint!(0));
+    assert!(point.1.clone() % prime != bigint!(0_i32));
```
Comment (Oppen) [inline src/math_utils.rs:139]:
```suggestion
    assert!(!(point.1.clone() % prime).is_zero());
```

File: src/vm/hints/secp/field_utils.rs (hunk 5 above)
Comment (Oppen) [reply src/vm/hints/secp/field_utils.rs:69]: I mean the `&*SECP_P`. Wouldn't it suffice to do `SECP_P`?

File: src/vm/hints/secp/signature.rs (hunk 8 above)
Comment (Oppen) [reply src/vm/hints/secp/signature.rs:53]: I hadn't observed that when using it before. Did you try it and it failed?

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/signature.rs (hunk 8 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/signature.rs:53]: Yes, and the best solution I've found is tu use the `&*`

File: src/math_utils.rs (hunk 16 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/c4555771386e2ddf020d6b63470f572ee5c938f3

File: src/math_utils.rs (hunk 15 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/c4555771386e2ddf020d6b63470f572ee5c938f3

File: src/math_utils.rs (hunk 14 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/c4555771386e2ddf020d6b63470f572ee5c938f3

File: src/math_utils.rs (hunk 13 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/c4555771386e2ddf020d6b63470f572ee5c938f3

File: src/math_utils.rs (hunk 12 above)
Comment (Juan-M-V) [reply src/math_utils.rs]: resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/c4555771386e2ddf020d6b63470f572ee5c938f3

File: src/vm/hints/secp/field_utils.rs (hunk 5 above)
Comment (Juan-M-V) [reply src/vm/hints/secp/field_utils.rs]: Nvm: https://github.com/lambdaclass/cleopatra_cairo/pull/333/commits/8008c87b35849d56aa208930c471b2b5290e54fc

File: src/vm/hints/secp/signature.rs [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+    vm.exec_scopes
+        .assign_or_update_variable("res", PyValueType::BigInt(value));
+    Ok(())
+}
+
+// Implements hint:
+// value = k = safe_div(res * b - a, N)
+pub fn div_mod_n_safe_div(vm: &mut VirtualMachine) -> Result<(), VirtualMachineError> {
+    let a = get_int_from_scope_ref(vm, "a")?.clone();
+    let b = get_int_from_scope_ref(vm, "b")?.clone();
```
Comment (Oppen) [inline src/vm/hints/secp/signature.rs:50]: You should be able to work with the refs here.

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/signature.rs [hunk 19]
```diff
@@ -57,6 +57,25 @@ pub fn div_mod_n_safe_div(vm: &mut VirtualMachine) -> Result<(), VirtualMachineE
...
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let x_cube_int = pack_from_var_name("x_cube", ids, vm, hint_ap_tracking)? % &*SECP_P;
+    let y_cube_int = (x_cube_int + &*BETA) % &*SECP_P;
+    let mut y = y_cube_int.modpow(&((&*SECP_P + 1) / 4), &*SECP_P);
+
+    let v = get_integer_from_var_name("v", ids, vm, hint_ap_tracking)?;
+    if v % 2 != &y % 2 {
```
Comment (Oppen) [inline src/vm/hints/secp/signature.rs:70]:
```suggestion
    if v & 1 != &y & 1 {
```

**Review by Oppen — CHANGES_REQUESTED** [review]
